documentation/dataplane/dash-byo-data-plane-app.md

```diff
@@ -0,0 +1,206 @@
+   2. [2.2. Initialization work flow](#22-initialization-work-flow)
+   3. [2.3. ASIC programming work flow](#23-asic-programming-work-flow)
+   4. [2.4. Flow management](#24-flow-management)
+   5. [2.5. RSS support](#25-rss-support)
+3. [3. SAI API design](#3-sai-api-design)
+   1. [3.1. SAI switch attribute updates](#31-sai-switch-attribute-updates)
+   2. [3.2. DASH pipeline programming APIs](#32-dash-pipeline-programming-apis)
+4. [4. Future work](#4-future-work)
+
+## 1. Data plane app
+
+Data plane app is one of the most important pieces in DASH pipeline. It is mainly responsible for:
+
+- Flow management, such as flow creation, deletion, resimulation, etc.
+- Work with ASIC to complete the table lookup and packet transformations. Depends on ASIC capability, if any operation could not be completed inside ASIC, it can be postponed to data plane app and completed in software.
+- ASIC management, such as initializing and programming match stage, etc.
+- and more.
+
+![DASH data plane app](./images/dash-default-data-plane-app.svg)
+
+By default, DASH technology providers will provide their own data plane app. This is usually closed source and can satisfy the DASH community requirement. However, in some cases, you might want to use your own data plane app due to some reasons, such as: private protocol support or support special logic in packet handling that is not supported by DASH technology providers. In this case, you can enable DASH BYO data plane app.
+
+## 2. BYO data plane app
+
+BYO data plane app is essentially a DPDK application that directly interacts with the data plane netdev.
+
+When BYO data plane app is enabled, the data plane app provided by technology providers will be either disabled or running without tied to the data plane netdev and only act as part of SAI API implementation if needed. With this setup, DASH users can start to use that device to run their own data plane app and process the packets.
+
+### 2.1. DASH management role and system overview
+
+With BYO data plane app, from our customer's perspective, we will have 2 sources to program the ASIC, such as creating match stage entries or managing flow entries:
+
+1. DASH users can explicitly program the ASIC via DASH SAI API proxy. A typical scenario is programming a new SDN policy.
+2. BYO data plane app can also program the ASIC. A typical scenario is creating flow entries.
+
+Even further, these 2 sources sometimes might need to work together in certain scenarios. For example, when a VNET mapping is updated, we need to update the VNET mapping entry as well as triggering flow resimulation to update all the flow entries that are related to this VNET mapping.
+
+Hence we need a design to avoid the same set of SAI APIs being called in 2 different processes accidentally and causing problems, such as managing the stage entries, as the last caller will overwrite the ASIC state without any synchronization and knowledge from the other side.
```

Comment:
I think this can be resolved by capability query and enabling only one path for certain set of APIs.

Reply:
yea, will update the doc and remove the proxy, which is actually my first version of design: https://github.com/r12f/DASH/blob/9d3c8ec2d9bc255ca72328cdce24a7fe9a037ac5/documentation/dataplane/dash-byo-data-plane-app.md.

XD